sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/ApproximateDistinct.java

```diff
@@ -0,0 +1,533 @@
+      return input.apply(Combine.globally(ApproximateDistinctFn
+              .<InputT>create(MoreObjects.firstNonNull(this.inputCoder(), input.getCoder()))
+              .withPrecision(this.precision())
+              .withSparseRepresentation(this.sparsePrecision())));
+    }
+  }
+
+  /**
+   * Implementation of {@link #perKey()}.
+   *
+   * @param <K>       type of the keys mapping the elements
+   * @param <V>       type of the values being combined per key
+   */
+  @AutoValue
+  public abstract static class PerKeyDistinct<K, V>
+          extends PTransform<PCollection<KV<K, V>>, PCollection<KV<K, HyperLogLogPlus>>> {
+
+    @Nullable abstract Coder<V> inputCoder();
+    abstract int precision();
+    abstract int sparsePrecision();
+    abstract Builder<K, V> toBuilder();
+
+    static <K, V> Builder<K, V> builder() {
+      return new AutoValue_ApproximateDistinct_PerKeyDistinct.Builder<K, V>()
+              .setInputCoder(null)
+              .setPrecision(12)
+              .setSparsePrecision(0);
+    }
+
+    @AutoValue.Builder
+    abstract static class Builder<K, V> {
+      abstract Builder<K, V> setInputCoder(Coder<V> coder);
+      abstract Builder<K, V> setPrecision(int p);
+      abstract Builder<K, V> setSparsePrecision(int sp);
+      abstract PerKeyDistinct<K, V> build();
+    }
+
+    public PerKeyDistinct<K, V> withInputCoder(Coder<V> coder) {
+      return toBuilder().setInputCoder(coder).build();
+    }
+
+    public PerKeyDistinct<K, V> withPrecision(int p) {
+      return toBuilder().setPrecision(p).build();
+    }
+
+    public PerKeyDistinct<K, V> withSparsePrecision(int sp) {
+      return toBuilder().setSparsePrecision(sp).build();
+    }
+
+    @Override
+    public PCollection<KV<K, HyperLogLogPlus>> expand(PCollection<KV<K, V>> input) {
+      KvCoder<K, V> inputCoder = (KvCoder<K, V>) input.getCoder();
+      return input.apply(Combine.<K, V, HyperLogLogPlus>perKey(ApproximateDistinctFn
+              .<V>create(MoreObjects.firstNonNull(this.inputCoder(), inputCoder.getValueCoder()))
+              .withPrecision(this.precision())
+              .withSparseRepresentation(this.sparsePrecision())));
+    }
+  }
+
+  /**
+   * A {@link CombineFn} that computes the stream into a {@link HyperLogLogPlus}
+   * sketch. Can be used as a state cell using stateful transforms such as
+   * {@link org.apache.beam.sdk.transforms.ParDo}s.
+   *
+   * <p>For more information about the parameters, see {@link ApproximateDistinct}.
+   *
+   * <h2>Examples using the {@link CombineFn}</h2>
+   *
+   * <p>By default, one must always specify a coder, using the {@link #create(Coder)} method.
+   * <br>By default, the precision is set to {@code 12} for a relative error of around {@code 2%}.
+   * <br>By default, the sparse representation is not used. One should use it if the cardinality
+   * may be less than {@code 12000}.
+   *
+   * <h4>Example 1 : default use</h4>
```

Comment:
All this documentation seems to almost mirror the one at class level. Could you deduplicate the documentation - eg. explain all the concepts at class level, and here just say that this implements the combine fn?

Reply:
Yes I can do that. I will put the references at the beginning of the comments so I am sure that the readers will see the links.